lib/src/css_mirroring/bidi_css_gen.dart

```diff
@@ -0,0 +1,355 @@
+        }
+      }
+      else {
+        checkState(originalTopLevels[iTopLevel].runtimeType == flippedTopLevels[iTopLevel].runtimeType);
+      }
+    }
+  }
+
+  _appendDirectionToRuleSet(_PendingRemovals removals, RuleSet ruleSet, Direction targetDirection) {
+    removals._transaction.edit(ruleSet.span.start.offset, ruleSet.span.end.offset,
+        ':host-context([dir="${enumName(targetDirection)}"]) ' +
+            ruleSet.span.text);
+  }
+
+  _removeStoredRules(_PendingRemovals removals, List<int> ruleStartPoints, List<int> ruleEndPoints) {
+    for(int iPoint = 0; iPoint < ruleStartPoints.length; iPoint++) {
+      removals._transaction.edit(ruleStartPoints[iPoint], ruleEndPoints[iPoint], '');
+    }
+  }
+  /// Removes a rule from the transaction.
+  _removeRuleSet(_PendingRemovals removals, int iTopLevel) {
+    removals._transaction.edit(_getRuleSetStart(removals._topLevels[iTopLevel]), _getRuleSetEnd(removals._topLevels, iTopLevel, removals.source.length), '');
+  }
+
+  _removeDirective(_PendingRemovals removals, TreeNode topLevel) {
+    removals._transaction.edit(_getRuleSetStart(topLevel) , topLevel.span.end.offset, '');
+  }
+
+  static int _getRuleSetStart(TreeNode node) {
+    if(node is RuleSet)
+      return node.span.start.offset;
+    // In case of Directives since the node span start does not include '@' so additional -1 is required.
+    return node.span.start.offset - 1;
+  }
+
+  static int _getRuleSetEnd(List<RuleSet> ruleSets, int iTopLevel, int parentEnd) {
+    var end = iTopLevel < ruleSets.length - 1
+        ? _getRuleSetStart(ruleSets[iTopLevel + 1])
+        : parentEnd;
+    return end;
+  }
+
+  static int _getDeclarationEnd(String source, int iDecl, var originalDecls, var decls) {
+    if (iDecl < originalDecls.length - 1) {
+      return decls[iDecl + 1].span.start.offset;
+    } else {
+      int fileLength = source.length;
+      int fromIndex = decls[iDecl].span.end.offset;
+      try {
+        while (fromIndex + 1 < fileLength) {
+          if (source.substring(fromIndex, fromIndex + 2) == '/*') {
+            while (source.substring(fromIndex, fromIndex + 2) != '*/') {
+              fromIndex++;
+            }
+          }
+          else if (source[fromIndex + 1] == '}')
+            return fromIndex + 1;
+          fromIndex++;
+        }
+      }
+      catch (exception, stackTrace) {
+        print('Invalid Css');
+      }
+    }
+  }
+
+  bool _isRuleRemovable(_PendingRemovals removals, var originalTopLevel) =>
+      removals._declStartPoints.length == originalTopLevel.declarationGroup.declarations.length;
+
+  /// Checks if a topLevel is direction independent.
+  bool _isDirectionIndependent(var originalTopLevel) {
+    List _directionIndependentDirectives = [
+      CharsetDirective,
+      FontFaceDirective,
+      ImportDirective,
+      KeyFrameDirective,
+      NamespaceDirective
+    ];
+    if(_directionIndependentDirectives.contains(originalTopLevel.runtimeType))
```

Comment:
Mmh, testing exact type is tricky: if suddenly csslib decides to have two subclasses of ImportDirective, this will break horribly.

Instead, you can return node is CharSetDirective || node is FontFaceDirective || ...